full-moon/src/ast/parsers.rs

```diff
@@ -2437,21 +2449,62 @@ fn parse_simple_type(
 #[cfg(feature = "luau")]
 fn parse_type_suffix(
     state: &mut ParserState,
-    simple_type: ast::TypeInfo,
+    simple_type: Option<ast::TypeInfo>,
 ) -> ParserResult<ast::TypeInfo> {
     let mut is_union = false;
     let mut is_intersection = false;
 
+    let mut types = Punctuated::new();
+
+    let leading = if simple_type.is_some() {
+        None
+    } else {
+        let current_token = match state.current() {
+            Ok(token) => token,
+            Err(()) => {
+                unreachable!("parse_type_suffix called with no simple_type and no current token")
+            }
+        };
+
+        match current_token.token_type() {
+            TokenType::Symbol {
+                symbol: Symbol::Pipe,
+            } => is_union = true,
+
+            TokenType::Symbol {
+                symbol: Symbol::Ampersand,
+            } => is_intersection = true,
+
+            _ => unreachable!("parse_type_suffix called with no simple_type and no `|` or `&`"),
+        }
+
+        Some(state.consume().unwrap())
+    };
+
     let mut current_type = simple_type;
 
     loop {
-        let Ok(current_token) = state.current() else {
-            return ParserResult::NotFound;
+        let ty = if current_type.is_some() {
```

Comment:
`.take().unwrap_or_else(`

Reply:
this doesn't work because of the `return`